Mathlib/RingTheory/Polynomial/Eisenstein/Generalized.lean

```diff
@@ -0,0 +1,169 @@
+  and such that the image `f` in `K[X]` is a power of `q`.
+  Assume moreover that `f.modByMonic q` is not zero in `(R ⧸ (P ^ 2))[X]`,
+  where `P` is the kernel of `algebraMap R K`.
+  Then `f` is irreducible. -/
+theorem generalizedEisenstein {q f : R[X]} {p : ℕ}
+    (hq_irr : Prime (q.map (algebraMap R K))) (hq_monic : q.Monic)
+    (hfd0 : 0 < natDegree f) (hf_monic : f.Monic)
+    (hfmodP : f.map (algebraMap R K) = q.map (algebraMap R K) ^ p)
+    (hfmodP2 : (f.modByMonic q).map (mk ((ker (algebraMap R K)) ^ 2)) ≠ 0) :
+    Irreducible f where
+  not_unit := mt degree_eq_zero_of_isUnit fun h => by
+    simp_all [lt_irrefl, natDegree_pos_iff_degree_pos]
+  isUnit_or_isUnit' g h h_eq := by
+    -- We have to show that factorizations `f = g * h` are trivial
+    set P : Ideal R := ker (algebraMap R K)
+    have hP : P.IsPrime := ker_isPrime (algebraMap R K)
+    obtain ⟨m, r, hg, hr, hm0⟩ :=
+      generalizedEisenstein_aux hq_irr hq_monic hf_monic hfmodP (h_eq ▸ dvd_mul_right g h)
+    obtain ⟨n, s, hh, hs, hn0⟩ :=
+      generalizedEisenstein_aux hq_irr hq_monic hf_monic hfmodP (h_eq ▸ dvd_mul_left h g)
+    by_cases hm : m = 0
+    -- If `m = 0`, `generalizedEisenstein_aux` shows that `g` is a unit.
+    · left; exact hm0 hm
+    by_cases hn : n = 0
+    -- If `n = 0`, `generalizedEisenstein_aux` shows that `h` is a unit.
+    · right; exact hn0 hn
+    -- Otherwise, we will get a contradiction by showing that `f %ₘ q` is zero mod `P ^ 2`.
+    exfalso
+    apply hfmodP2
+    suffices f %ₘ q = (r * s) %ₘ q by
+      -- Since the coefficients of `r` and `s` are in `P`, those of `r * s` are in `P ^ 2`
+      simp only [ext_iff, ← coeff_map, eq_zero_iff_mem]
+      rw [← ext_iff, this, map_modByMonic _ hq_monic]
+      convert zero_modByMonic _
+      ext n
+      rw [coeff_map, coeff_mul, map_sum, coeff_zero]
+      apply Finset.sum_eq_zero
+      intro x hx
+      rw [eq_zero_iff_mem, pow_two]
+      apply mul_mem_mul
+      · rw [mem_ker, ← coeff_map, hr, coeff_zero]
+      · rw [mem_ker, ← coeff_map, hs, coeff_zero]
+    -- It remains to prove the equality `f %ₘ q = (r * s) %ₘ q`, which is straightforward
+    rw [h_eq, hg, hh]
+    simp only [add_mul, mul_add, map_add, ← modByMonicHom_apply]
+    simp only [← add_assoc, modByMonicHom_apply]
+    convert zero_add _
+    convert zero_add _
+    · convert zero_add _
+      · rw [modByMonic_eq_zero_iff_dvd hq_monic]
+        exact ((dvd_pow_self q hn).mul_left _).mul_left _
+      · symm
+        rw [modByMonic_eq_zero_iff_dvd hq_monic]
+        simp only [← mul_assoc]
+        exact (dvd_pow_self q hn).mul_left _
+    · symm
+      rw [modByMonic_eq_zero_iff_dvd hq_monic]
+      exact ((dvd_pow_self q hm).mul_left _).mul_right _
+
+-- move
+theorem CharP.ker_intAlgebraMap_eq_span
+    {R : Type*} [Ring R] (p : ℕ) [CharP R p] :
+    RingHom.ker (algebraMap ℤ R) = Ideal.span {(p : ℤ)} := by
+  ext a
+  simp [CharP.intCast_eq_zero_iff R p, Ideal.mem_span_singleton]
```

Comment:
Does `#find_home! CharP.ker_intAlgebraMap_eq_span` give any sensible suggestions?

Reply:
no, nothing.

Reply:
but I searched for a lemma (using loogle) that has `CharP` and `Ideal.span`,  and I was led to `Mathlib/Algebra/CharP/Quotient.lean` where the lemma fits.